rehearse: use updateconfig to generate CMs

bbguimaraes opened the pull request:
I am still working on templates, as that will require more fundamental changes, but the basic mechanism is already reviewable.

bbguimaraes committed:
rehearse: use updateconfig to generate CMs

stevekuznetsov commented on pkg/config/template.go:
```diff
@@ -26,6 +26,11 @@ import (
 	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	prowgithub "k8s.io/test-infra/prow/github"
+	_ "k8s.io/test-infra/prow/hook"
```
I should add a non-validating load so we don't need this -.-

bbguimaraes replied:
Does the configuration check we have also validate plugins? If not (and even if it is?), validating this may be an additional benefit of `rehearse`.

stevekuznetsov replied:
It does, yeah

stevekuznetsov reviewed (commented):
@bbguimaraes you want to make sure this is ready to go for the profiles and then do templates in another PR?

bbguimaraes commented:
I am considering it. The template changes are almost done, they would make the PR significantly larger and are mostly orthogonal.

It would be nice to get #122 in before this so I can add proper integration tests, I'll take a look and see what's left in there.

petr-muller commented:
LGTM, nice! Would prefer to merge this and have another PR for templates...

bbguimaraes commented:
Rebased and updated. If you've reviewed this before, I tried to keep the diff to a minimum: check only the last commit. I'll squash when approved. This can be merged as is, but you can also check the template PR:

https://github.com/openshift/ci-operator-prowgen/pull/146

Main differences from the previous version:
- fixed path in `main.go` (caught by https://github.com/openshift/ci-operator-prowgen/pull/140)
- fixed to create only necessary profiles (caught by https://github.com/openshift/ci-operator-prowgen/pull/143)
- added necessary files for the new integration tests in https://github.com/openshift/ci-operator-prowgen/pull/140
- added an `update` reactor to the dry-run client, the `create` reactor is kept because of the old template code and is removed in the next PR

stevekuznetsov commented on pkg/config/template.go:
```diff
@@ -103,43 +108,53 @@ func (c *TemplateCMManager) CreateCMTemplates(templates CiTemplates) error {
+					Status:   string(prowgithub.PullRequestFileModified),
+				})
+			}
 			return err
```
It's impossible for `filepath.Rel` to error here right? Might be worth a comment

bbguimaraes replied:
Added.

stevekuznetsov commented on pkg/config/template.go:
```diff
@@ -103,43 +108,53 @@ func (c *TemplateCMManager) CreateCMTemplates(templates CiTemplates) error {
+				break
+			}
+		}
+		err := c.createCM(&v1.ConfigMap{
```
Update should do creation, why do we need this?

bbguimaraes replied:
To add the labels. This is something we could easily add to the upstream library in the future.

stevekuznetsov replied:
Can we rely on upstream to create and just add labels ourselves after?

bbguimaraes replied:
That was the other option, but it would require a `Get` + `Update` here (`updateconfig.Update` doesn't return what it created), while `updateconfig` already does a `Get` first.

bbguimaraes replied:
For reference: https://github.com/kubernetes/test-infra/blob/50de43197c2fdf95570eb414e3510a0b562704f9/prow/plugins/updateconfig/updateconfig.go#L90

bbguimaraes replied:
Doing that also leaves a (very unlikely, but still…) window where we could create CMs that would never be removed if we failed before labeling them.

bbguimaraes commented:
Incorporated @petr-muller's [suggestion](https://github.com/openshift/ci-operator-prowgen/pull/146#discussion_r276610780) and added an extra test fix that got lost during the refactorings.

bbguimaraes commented:
Rebase and squashed.

Some small changes were necessary because of an upstream change to the interface of the `updateconfig` functions: https://github.com/kubernetes/test-infra/commit/8ca35465a98da0c769756bc75e9e00fdc851001d.

stevekuznetsov reviewed (approved):
/lgtm
/approve
